### PR TITLE
Dft Part 5

### PR DESCRIPTION
**Pull request by GreatRSingh**
## Description

Fix #(issue)

## Type of change

Please check the option that is related to your PR.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
  - In this case, we recommend to discuss your modification on GitHub issues before creating the PR
- [ ] Documentations (modification for documents)

## Checklist

- [x] My code follows [the style guidelines of this project](https://deepchem.readthedocs.io/en/latest/development_guide/coding.html)
  - [x] Run `yapf -i <modified file>` and check no errors (**yapf version must be  0.32.0**)
  - [x] Run `mypy -p deepchem` and check no errors
  - [x] Run `flake8 <modified file> --count` and check no errors
  - [x] Run `python -m doctest <modified file>` and check no errors
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New unit tests pass locally with my changes
- [x] I have checked my code and corrected any misspellings

### CODE REVIEW

**Comment by shreyasvinaya** [comment]
@GreatRSingh can you please rebase, you should see a lot more tests pass

**Comment by GreatRSingh** [comment]
I have not added the test for hf class because it was requiring a lot of parameters and other objects which could have made the test and example very long, if its ok then I will add its test and example after writing the Mol class.

File: deepchem/utils/dft_utils/data/datastruct.py [hunk 1]
```diff
@@ -58,15 +58,30 @@ def __repr__(self) -> str:
...
-
-        return fcn(self.u, self.d)
+        if isinstance(a, SpinParam):
+            return fcn(a.u, a.d)
+        else:
+            return a
+
+    @staticmethod
+    def apply_fcn(fcn: Callable[..., P], *a):
+        """"apply the function for each up and down elements of a"""
```
Comment (rbharath) [inline deepchem/utils/dft_utils/data/datastruct.py:77]: Nit: capitalization "Apply". Please fix in a follow up PR

File: deepchem/utils/dft_utils/qccalc/hf.py [hunk 2]
```diff
@@ -0,0 +1,505 @@
...
+                                    dm_penalty))
+        else:
+            dm = dm_penalty
+            penalty = None
+        return dm, penalty
+
+    def pack_aoparams(
+            self, aoparams: Union[torch.Tensor,
+                                  SpinParam[torch.Tensor]]) -> torch.Tensor:
+        """if polarized, then pack it by concatenating them in the last dimension
```
Comment (rbharath) [inline deepchem/utils/dft_utils/qccalc/hf.py:267]: Capitalization

File: deepchem/utils/dft_utils/qccalc/scf_qccalc.py [hunk 3]
```diff
@@ -0,0 +1,595 @@
+from __future__ import annotations
```
Comment (rbharath) [inline deepchem/utils/dft_utils/qccalc/scf_qccalc.py:1]: Add a link to the original repo to give them credit

File: deepchem/utils/dft_utils/qccalc/scf_qccalc.py [hunk 4]
```diff
@@ -0,0 +1,595 @@
...
+                pc = SpinParam.apply_fcn(
+                    lambda dm, norb: h.dm2ao_orb_params(SpinParam.sum(dm),
+                                                        norb=norb), dm, norb)
+                p = SpinParam.apply_fcn(lambda pc: pc[0], pc)
+                c = SpinParam.apply_fcn(lambda pc: pc[1], pc)
+                params = self._engine.pack_aoparams(p)
+                coeffs = self._engine.pack_aoparams(c)
+                return params, coeffs
+
+            def params2dm(params: torch.Tensor, coeffs: torch.Tensor) \
```
Comment (rbharath) [inline deepchem/utils/dft_utils/qccalc/scf_qccalc.py:179]: Docstring

File: deepchem/utils/test/test_dft_utils.py [hunk 5]
```diff
@@ -506,3 +506,46 @@ def test_get_grid_transform():
...
+                                SpinParam[torch.Tensor]]) -> torch.Tensor:
+            if isinstance(dm, SpinParam):
+                return dm.u + dm.d * 1.1
+            return dm * 1.1
+
+    myEngine = engine()
+    a = SCF_QCCalc(myEngine)
+    assert torch.allclose(a.dm2energy(torch.tensor([1.1])),
+                          torch.tensor([1.2100]))
+
```
Comment (rbharath) [inline deepchem/utils/test/test_dft_utils.py:532]: We are missing tests for a lot of functionality. Please add

File: deepchem/utils/dft_utils/qccalc/hf.py [hunk 6]
```diff
@@ -0,0 +1,505 @@
...
+
+        engine = HFEngine(system, restricted)
+        super().__init__(engine, variational)
+
+
+class HFEngine(BaseSCFEngine):
+    """
+    Engine to be used with Hartree Fock.
+    This class provides the calculation of the self-consistency iteration step
+    and the calculation of the post-calculation properties.
```
[the remark this answers is not in the document]
Comment (GreatRSingh) [reply deepchem/utils/dft_utils/qccalc/hf.py:40]: Done

File: deepchem/utils/dft_utils/qccalc/hf.py [hunk 7]
```diff
@@ -0,0 +1,505 @@
...
+
+        """
+        eigvals, eigvecs = self.diagonalize(fock, self._norb)
+        dm = SpinParam.apply_fcn(
+            lambda eivecs, orb_weights: self._hamilton.ao_orb2dm(
+                eivecs, orb_weights), eigvecs, self._orb_weight)
+        return dm
+
+    def diagonalize(self, fock: Union[LinearOperator, SpinParam[LinearOperator]], norb: Union[int, SpinParam[int]]) -> \
+            Union[Tuple[torch.Tensor, torch.Tensor], Tuple[SpinParam[torch.Tensor], SpinParam[torch.Tensor]]]:
```
[the remark this answers is not in the document]
Comment (GreatRSingh) [reply deepchem/utils/dft_utils/qccalc/hf.py:497]: Done

File: deepchem/utils/dft_utils/qccalc/scf_qccalc.py (hunk 3 above)
Comment (GreatRSingh) [reply deepchem/utils/dft_utils/qccalc/scf_qccalc.py:1]: done

File: deepchem/utils/dft_utils/qccalc/scf_qccalc.py [hunk 8]
```diff
@@ -0,0 +1,595 @@
...
+            # post-process parameters
+            self._dm = self._engine.scp2dm(scp)
+        else:
+            system = self.get_system()
+            h = system.get_hamiltonian()
+            orb_weights = system.get_orbweight(polarized=self._polarized)
+            norb = SpinParam.apply_fcn(lambda orb_weights: len(orb_weights),
+                                       orb_weights)
+
+            def dm2params(dm: Union[torch.Tensor, SpinParam[torch.Tensor]]) -> \
```
[the remark this answers is not in the document]
Comment (GreatRSingh) [reply deepchem/utils/dft_utils/qccalc/scf_qccalc.py:168]: Done.
They were actually inside a function and got ignored earlier.

File: deepchem/utils/dft_utils/qccalc/scf_qccalc.py (hunk 4 above)
Comment (GreatRSingh) [reply deepchem/utils/dft_utils/qccalc/scf_qccalc.py:179]: Done

File: deepchem/utils/test/test_dft_utils.py (hunk 5 above)
Comment (GreatRSingh) [reply deepchem/utils/test/test_dft_utils.py:532]: Added more tests. It's a complicated base class, tested many things in its sub class. Will add more tests after pylibxc and its API functions get implemented.

**Review by rbharath — COMMENTED** [review]
@shreyasvinaya Can you do a quick review for a CI check?

**Review by rbharath — APPROVED** [review]
LGTM

**Pull request merged**